src/main/java/de/zalando/aruha/nakadi/service/EventStreamCoordinator.java

```diff
@@ -0,0 +1,174 @@
+        this.topicRepository = topicRepository;
+        eventStreams = Collections.synchronizedList(Lists.newArrayList());
+        currentTopologies = Collections.synchronizedMap(Maps.newHashMap()); // todo: we need somehow to lock it between reading and writing
+    }
+
+    /**
+     * Creates new event stream and invokes clients rebalancing
+     *
+     * @param subscriptionId the id of subscription
+     * @param outputStream the output stream where to stream data
+     * @param config the configuration of stream
+     * @return event stream created
+     */
+    public EventStream createEventStream(final String subscriptionId, final OutputStream outputStream,
+                                         final EventStreamConfig config) {
+        // add new client to topology and invoke rebalance
+        final String newClientId = subscriptionRepository.generateNewClientId();
+        subscriptionRepository.addClient(subscriptionId, newClientId);
+        rebalanceSubscriptionIfNeeded(subscriptionId);
+
+        // get partitions and cursors for this new client
+        final Topology topology = subscriptionRepository.getTopology(subscriptionId);
+        final int newClientIndex = topology.getClientIndex(newClientId).get();
+        final int clientsNum = topology.getClientIds().size();
+
+        final List<TopicPartition> partitionsForClient = partitionDistributor.getPartitionsForClients(subscriptionId,
+                ImmutableList.of(newClientIndex), clientsNum).get(newClientIndex);
+
+        final List<Cursor> cursors = partitionsForClient
+                .stream()
+                .map(tp -> subscriptionRepository.getCursor(subscriptionId, tp.getTopic(), tp.getPartition()))
+                .collect(Collectors.toList());
+
+        // create the event stream itself
+        final EventConsumer eventConsumer = topicRepository.createEventConsumer(cursors);
+        final EventStream eventStream = new EventStream(eventConsumer, outputStream, config, cursors);
+        eventStream.setClientId(newClientId);
+        eventStream.setSubscriptionId(subscriptionId);
+        eventStreams.add(eventStream);
+        return eventStream;
+    }
+
+    /**
+     * Removes finished stream and invokes rebalance of this stream subscription
+     *
+     * @param eventStream the stream to remove
+     */
+    public void removeEventStream(final EventStream eventStream) {
+        eventStreams.remove(eventStream);
+        subscriptionRepository.removeClient(eventStream.getSubscriptionId(), eventStream.getClientId());
+        rebalanceSubscriptionIfNeeded(eventStream.getSubscriptionId());
+    }
+
+    /**
+     * Checks if rebalance is needed for subscriptions running on this Nakadi instance
+     * todo: in general this scheduled task should be probably replaced with watching appropriate zookeeper nodes
+     */
+    @Scheduled(fixedRate = SUBSCRIPTION_REBALANCE_CHECK_PERIOD_IN_MS)
+    private void rebalanceSubscriptions() {
```

Comment:
It will sound like it will rebalance ALL subscriptions if needed :-D But in fact it will rebalance only needed. Maybe `rebalanceNeededSubscriptions` ?